### PR TITLE
Cw 45 implement yat sending

### PR DESCRIPTION
**Pull request by godilite**
-Resolve YAT emoji address
-Update Sending UI for YAT

### CODE REVIEW

File: lib/entities/yat_record.dart [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      results.forEach((dynamic key, dynamic value) {
+        yatRecords.add(YatRecord.fromJson(value as Map<String, dynamic>));
+      });
+
+      return yatRecords;
+    } catch (_) {
+      return yatRecords;
+    }
+  }
+}
```
Comment (mkyq) [inline lib/entities/yat_record.dart:54]: Please write another one class for fetching addresses by emoji id. For example you can make `YatService` class which will fetch address by emoji id by using regular (non static) method.

File: lib/entities/yat_record.dart [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  YatRecord.fromJson(Map<String, dynamic> json) {
+    address = json['address'] as String;
+    category = json['category'] as String;
+  }
+
+  static const tags = {
+    "XMR": '0x1001,0x1002',
+    "BTC": '0x1003',
+    "LTC": '0x3fff'
```
Comment (mkyq) [inline lib/entities/yat_record.dart:28]: Prefer to use single quotes

File: lib/entities/yat_record.dart [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import 'dart:convert';
+
+import 'package:http/http.dart';
+
+/// Resolves YAT record
+class YatRecord {
+  String category;
+  String address;
+
+  static const yatBaseUrl = "https://a.y.at";
```
Comment (mkyq) [inline lib/entities/yat_record.dart:10]: Need to make sure that you can use testing url when its needed. For this please write `YatService` class where you could to set which url you need to use

File: lib/entities/parse_address_from_domain.dart [hunk 4]
```diff
@@ -26,15 +27,15 @@ Future<ParsedAddress> parseAddressFromDomain(
...
           return ParsedAddress(
               addresses: [domain], parseFrom: ParseFrom.yatRecord);
         }
 
         return ParsedAddress(
-            addresses: addresses, name: domain, parseFrom: ParseFrom.yatRecord);
+            addresses: addresses.map((e) => e.address).toList(), name: domain, parseFrom: ParseFrom.yatRecord);
       } catch (e) {
         return ParsedAddress(addresses: [domain]);
       }
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:41]: Let change that logic a bit. For example let check is user entered text emoji or not, then if it's emoji we will use Yat for resolving the address. If the text is not emoji let check is it domain addresses or not and will use openallias for resolve address.

**Review by mkyq — CHANGES_REQUESTED** [review]

File: lib/entities/yat_record.dart (hunk 2 above)
Comment (godilite) [reply lib/entities/yat_record.dart]: okay

File: lib/entities/yat_record.dart (hunk 3 above)
Comment (godilite) [reply lib/entities/yat_record.dart]: done

File: lib/entities/parse_address_from_domain.dart (hunk 4 above)
Comment (godilite) [reply lib/entities/parse_address_from_domain.dart]: done

File: lib/entities/yat_record.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  YatRecord.fromJson(Map<String, dynamic> json) {
+    address = json['address'] as String;
+    category = json['category'] as String;
+  }
+
+  static const tags = {
+    'XMR': '0x1001,0x1002',
+    'BTC': '0x1003',
+    'LTC': '0x3fff'
+  };
```
Comment (mkyq) [inline lib/entities/yat_record.dart:19]: Let move that to YatService

File: lib/entities/parse_address_from_domain.dart [hunk 6]
```diff
@@ -24,17 +25,21 @@ Future<ParsedAddress> parseAddressFromDomain(
...
+          final addresses = await yatService.fetchYatAddress(domain, ticker);
+          if(addresses?.isEmpty ?? true){
+            return ParsedAddress(addresses: [domain], parseFrom: ParseFrom.yatRecord);
+          }
           return ParsedAddress(
-              addresses: [domain], parseFrom: ParseFrom.yatRecord);
+              addresses: addresses.map((e) => e.address).toList(),
+              name: domain,
+              parseFrom: ParseFrom.yatRecord);
         }
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:41]: Move that part above before `OpenaliasRecord.formatDomainName(domain);` calling. Let firstly decide is it emoji or domain and then call needed instructions.

**Review by mkyq — CHANGES_REQUESTED** [review]

**Comment by mkyq** [comment]
Also please rework `parse_address_from_domain.dart`. Let make that as class which will has `parse(String text, String ticker)` method. And please firstly need to decide what text type is (emoji, or domain) and then call needed resolver.

**Comment by mkyq** [comment]
cc @godilite

File: lib/entities/parse_address_from_domain.dart [hunk 7]
```diff
@@ -17,57 +20,60 @@ const unstoppableDomains = [
...
-    final domainParts = formattedName.split('.');
-    final name = domainParts.last;
-
-    if (domainParts.length <= 1 || domainParts.first.isEmpty || name.isEmpty) {
-      try {
-        final addresses = await fetchYatAddress(domain, ticker);
-
+Future<ParsedAddress> parse(String text, String ticker) async {
+    try {
+        if (text.hasOnlyEmojis) {
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:25]: Please add one tab back

File: lib/entities/parse_address_from_domain.dart [hunk 8]
```diff
@@ -17,57 +20,60 @@ const unstoppableDomains = [
...
         }
-
         return ParsedAddress(
-            addresses: addresses, name: domain, parseFrom: ParseFrom.yatRecord);
-      } catch (e) {
-        return ParsedAddress(addresses: [domain]);
+            addresses: addresses.map((e) => e.address).toList(),
+            name: text,
+            parseFrom: ParseFrom.yatRecord,
+          );
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:36]: Move to another one method like "ParsedAddress fetchEmojiAddress" or similar

File: lib/entities/parse_address_from_domain.dart [hunk 9]
```diff
@@ -17,57 +20,60 @@ const unstoppableDomains = [
...
+        if (address?.isEmpty ?? true) {
+          return ParsedAddress(addresses: [text]);
+        }
 
-      if (address?.isEmpty ?? true) {
-        return ParsedAddress(addresses: [domain]);
+        return ParsedAddress(
+            addresses: [address],
+            name: text,
+            parseFrom: ParseFrom.unstoppableDomains);
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:56]: Same here, please make another one method which will fetch address by domain and return parsed address

File: lib/entities/parse_address_from_domain.dart [hunk 10]
```diff
@@ -17,57 +20,60 @@ const unstoppableDomains = [
...
+        return ParsedAddress(addresses: [text]);
+      }
 
-    if (record == null || record.address.contains(formattedName)) {
-      return ParsedAddress(addresses: [domain]);
+      return ParsedAddress(
+          addresses: [record.address],
+          name: record.name,
+          description: record.description,
+          parseFrom: ParseFrom.openAlias);
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:70]: Same here, please make another one method which will fetch address by domain and return parsed address

File: lib/entities/parse_address_from_domain.dart [hunk 11]
```diff
@@ -17,57 +20,60 @@ const unstoppableDomains = [
...
-    final name = domainParts.last;
-
-    if (domainParts.length <= 1 || domainParts.first.isEmpty || name.isEmpty) {
-      try {
-        final addresses = await fetchYatAddress(domain, ticker);
-
+Future<ParsedAddress> parse(String text, String ticker) async {
+    try {
+        if (text.hasOnlyEmojis) {
+        final yatService = getIt.get<YatService>();
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:26]: Move YatService to the class property and inject it through constructor

File: lib/entities/parse_address_from_domain.dart [hunk 12]
```diff
@@ -1,10 +1,13 @@
...
+import 'package:cake_wallet/di.dart';
 import 'package:cake_wallet/entities/openalias_record.dart';
 import 'package:cake_wallet/entities/parsed_address.dart';
 import 'package:cake_wallet/entities/unstoppable_domain_address.dart';
-import 'package:cake_wallet/store/yat/yat_store.dart';
-import 'package:cw_core/wallet_type.dart';
+import 'package:cake_wallet/entities/emoji_string_extension.dart';
 
-const unstoppableDomains = [
+class ParseAddressFromDomain {
```
Comment (mkyq) [inline lib/entities/parse_address_from_domain.dart:8]: Rename class, as you parse not only from domain, i assume you can name it as "AddressResolver" and rename parse method to resolve.

**Review by mkyq — CHANGES_REQUESTED** [review]

**Comment by mkyq** [comment]
cc @godilite

File: lib/entities/parse_address_from_domain.dart (hunk 7 above)
Comment (godilite) [reply lib/entities/parse_address_from_domain.dart:25]: done

File: lib/entities/parse_address_from_domain.dart (hunk 11 above)
Comment (godilite) [reply lib/entities/parse_address_from_domain.dart:26]: done

File: lib/entities/parse_address_from_domain.dart (hunk 10 above)
Comment (godilite) [reply lib/entities/parse_address_from_domain.dart:70]: done

File: lib/entities/parse_address_from_domain.dart (hunk 9 above)
Comment (godilite) [reply lib/entities/parse_address_from_domain.dart:56]: done

File: lib/entities/parse_address_from_domain.dart (hunk 12 above)
Comment (godilite) [reply lib/entities/parse_address_from_domain.dart:8]: alright, done

**Review by mkyq — APPROVED** [review]

**Review by mkyq — APPROVED** [review]

**Pull request merged**